### DISCUSSION TITLE
Use of other graphviz layout engines?

### DISCUSSION THREAD

**Discussion by lenolib** [likec4/likec4#1142] [Ideas]
I'm really enjoying LikeC4! Very impressed, and compared to structurizr the vscode extension speeds up development so much.

I'm wondering if it would be relevant to support other layout engines supported by graphviz?

**Comment by davydkov** [comment]
Hey @lenolib 

> relevant to support other layout engines supported by graphviz?

I experimented with `fdp`.
For example, we have a diagram with manual layout (moved nodes/edges).
When model changed and new nodes need to placed - we want to apply manual positions and let graphviz layout new graph. Only `fdp` supports initial positions.

https://github.com/likec4/likec4/blob/f0a9ea696a8365d13803b6f4d2f79d9f38db1291/packages/layouts/src/graphviz/DotPrinter.ts#L434-L509

But I struggled to get edges with compounds, that why I have it:

https://github.com/likec4/likec4/blob/f0a9ea696a8365d13803b6f4d2f79d9f38db1291/packages/layouts/src/graphviz/GraphvizLayoter.ts#L49-L58

Do you know graphviz well?

**Reply by lenolib** [reply]
I see. Maintaining the manual positions when adding new ones does sound like a tricky constraint. Could maybe it be a manual layout option to "generate all new positions using engine <X>", and then for single new nodes, use dot/fdp with the existing positions as fixed inputs? (or is that what is used right now? sorry if I missunderstood).

I'm afraid I'm not knowledgable regarding graphqviz. The reason I posed the question is because I felt that the automatic layout for the architecture I'm expressing produced either very long edges or many edge crossings, and I felt curious what other algorithms would produce.

**Reply by lenolib** [reply]
I replaced the names of the nodes in the architecture I'm working on and compared the graphviz output and likec4. 
I got the subjectively best results using dot. LikeC4 seems to orient things more strongly top-bottom somehow which produces quite long edges.

_(note: In the real architecture, the gray lorem_ipsum_v node belongs to another context, so it is displayed outside of the rectangle containing the other nodes, but I don't know if or how that's expressible in graphviz, so I included it as just another node in this comparison)_

<img width="825" alt="image" src="https://github.com/user-attachments/assets/63c61126-9824-4d32-b12e-fd84e9b3394b">

<img width="1225" alt="image" src="https://github.com/user-attachments/assets/7eab7ef9-1d6c-4506-ac8e-b16d8917c606">

Graphviz pasted into https://magjac.com/graphviz-visual-editor/
```
digraph G {
    layout=dot
    rankdir=TB
    subgraph sub1 {
        node [shape=box]
        lorem_ipsum_a
        lorem_ipsum_b
        lorem_ipsum_c
        lorem_ipsum_d
        lorem_ipsum_e
        lorem_ipsum_f
        lorem_ipsum_g
        lorem_ipsum_h
        lorem_ipsum_i
        lorem_ipsum_j
        lorem_ipsum_k
        lorem_ipsum_l
        lorem_ipsum_m
        lorem_ipsum_n
        lorem_ipsum_o
        lorem_ipsum_p
        lorem_ipsum_q
        lorem_ipsum_r
        lorem_ipsum_s
        lorem_ipsum_t
        lorem_ipsum_u
        
        lorem_ipsum_a -> lorem_ipsum_b 
        lorem_ipsum_b -> lorem_ipsum_c 
        lorem_ipsum_d -> lorem_ipsum_e 
        lorem_ipsum_e -> lorem_ipsum_f 
        lorem_ipsum_g -> lorem_ipsum_h 
        lorem_ipsum_i -> lorem_ipsum_j
        lorem_ipsum_j -> lorem_ipsum_k
        lorem_ipsum_c -> lorem_ipsum_k
        lorem_ipsum_h -> lorem_ipsum_c 
        lorem_ipsum_l -> lorem_ipsum_c 
        lorem_ipsum_m -> lorem_ipsum_c 
        lorem_ipsum_m -> lorem_ipsum_d 
        lorem_ipsum_d -> lorem_ipsum_j 
        lorem_ipsum_d -> lorem_ipsum_n 
        lorem_ipsum_d -> lorem_ipsum_c 
        lorem_ipsum_d -> lorem_ipsum_o 
        lorem_ipsum_f -> lorem_ipsum_c 
        lorem_ipsum_p -> lorem_ipsum_c 
        lorem_ipsum_p -> lorem_ipsum_o 
        lorem_ipsum_q -> lorem_ipsum_c 
        lorem_ipsum_q -> lorem_ipsum_o 
        lorem_ipsum_d -> lorem_ipsum_r 
        lorem_ipsum_n -> lorem_ipsum_s 
        lorem_ipsum_s -> lorem_ipsum_t 
        lorem_ipsum_t -> lorem_ipsum_g 
        lorem_ipsum_u -> lorem_ipsum_f 
    }
    subgraph sub2 {
        node [shape=box style=filled, color=lightgray];
        lorem_ipsum_v
        
        lorem_ipsum_u -> lorem_ipsum_v
        lorem_ipsum_t -> lorem_ipsum_v
        lorem_ipsum_v -> lorem_ipsum_n
        lorem_ipsum_d -> lorem_ipsum_v
    }
}
```

LikeC4:
```
specification {
    element ctx
    element sys
}

model {
    the_context = ctx "The context"
    extend the_context {
    lorem_ipsum_v = sys "lorem_ipsum_v" {
        style {
            color gray
        }
    }
    lorem_ipsum_a = sys "lorem_ipsum_a"
    lorem_ipsum_b = sys "lorem_ipsum_b"
    lorem_ipsum_c = sys "lorem_ipsum_c"
    lorem_ipsum_d = sys "lorem_ipsum_d"
    lorem_ipsum_e = sys "lorem_ipsum_e"
    lorem_ipsum_f = sys "lorem_ipsum_f"
    lorem_ipsum_g = sys "lorem_ipsum_g"
    lorem_ipsum_h = sys "lorem_ipsum_h"
    lorem_ipsum_i = sys "lorem_ipsum_i"
    lorem_ipsum_j = sys "lorem_ipsum_j"
    lorem_ipsum_k = sys "lorem_ipsum_k"
    lorem_ipsum_l = sys "lorem_ipsum_l"
    lorem_ipsum_m = sys "lorem_ipsum_m"
    lorem_ipsum_n = sys "lorem_ipsum_n"
    lorem_ipsum_o = sys "lorem_ipsum_o"
    lorem_ipsum_p = sys "lorem_ipsum_p"
    lorem_ipsum_q = sys "lorem_ipsum_q"
    lorem_ipsum_r = sys "lorem_ipsum_r"
    lorem_ipsum_s = sys "lorem_ipsum_s"
    lorem_ipsum_t = sys "lorem_ipsum_t"
    lorem_ipsum_u = sys "lorem_ipsum_u"

    lorem_ipsum_a -> lorem_ipsum_b
    lorem_ipsum_b -> lorem_ipsum_c
    lorem_ipsum_d -> lorem_ipsum_e
    lorem_ipsum_e -> lorem_ipsum_f
    lorem_ipsum_g -> lorem_ipsum_h
    lorem_ipsum_i -> lorem_ipsum_j
    lorem_ipsum_j -> lorem_ipsum_k
    lorem_ipsum_c -> lorem_ipsum_k
    lorem_ipsum_h -> lorem_ipsum_c
    lorem_ipsum_l -> lorem_ipsum_c
    lorem_ipsum_m -> lorem_ipsum_c
    lorem_ipsum_m -> lorem_ipsum_d
    lorem_ipsum_d -> lorem_ipsum_j
    lorem_ipsum_d -> lorem_ipsum_n
    lorem_ipsum_d -> lorem_ipsum_c
    lorem_ipsum_d -> lorem_ipsum_o
    lorem_ipsum_f -> lorem_ipsum_c
    lorem_ipsum_p -> lorem_ipsum_c
    lorem_ipsum_p -> lorem_ipsum_o
    lorem_ipsum_q -> lorem_ipsum_c
    lorem_ipsum_q -> lorem_ipsum_o
    lorem_ipsum_d -> lorem_ipsum_r
    lorem_ipsum_n -> lorem_ipsum_s
    lorem_ipsum_s -> lorem_ipsum_t
    lorem_ipsum_t -> lorem_ipsum_g
    lorem_ipsum_u -> lorem_ipsum_f
    lorem_ipsum_u -> lorem_ipsum_v
    lorem_ipsum_t -> lorem_ipsum_v
    lorem_ipsum_v -> lorem_ipsum_n
    lorem_ipsum_d -> lorem_ipsum_v
    }

}

views {
    view of the_context {
        include *
	}
}
```

**Reply by lenolib** [reply]
Update: using the dot export I realized that the parameter value responsible for the main difference in layout is `TBbalance` which likec4 sets to `min`, whereas it's unset as default. Could it be an idea to add this as an option or to not set it?
ref: https://graphviz.org/docs/attrs/TBbalance/

Output with TBbalance none:
<img width="754" alt="image" src="https://github.com/user-attachments/assets/57ce3fb2-8a2b-4e69-ba16-55ea483b0cf2">
